File: sdks/java/turbine-sdk/src/main/java/io/turbine/sdk/Writer.java
package io.turbine.sdk;

import com.google.protobuf.ByteString;
import io.turbine.sdk.proto.AppendRequest;
import io.turbine.sdk.proto.AppendResponse;
import io.turbine.sdk.proto.AuthRequest;
import io.turbine.sdk.proto.BatchAck;
import io.turbine.sdk.proto.ClientMessage;
import io.turbine.sdk.proto.Record;
import io.turbine.sdk.proto.RecordBatch;
import io.turbine.sdk.proto.ServerMessage;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.nio.ByteBuffer;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Writer client for sending messages to Turbine.
 * <p>
 * Wire payloads use generated protobuf message classes directly.
 */
public class Writer implements AutoCloseable {
    private static final Logger log = LoggerFactory.getLogger(Writer.class);
    private static final int RESPONSE_QUEUE_CAPACITY = 8192;
    private static final int ERR_BACKPRESSURE = 1004;

    private final UUID writerId;
    private final AtomicLong seq = new AtomicLong(1);
    private final WriterConfig config;
    private final TurbineWebSocketClient client;
    private final BlockingQueue<byte[]> responseQueue;
    private final Map<Long, CompletableFuture<AppendResponse>> pendingResponses;
    private final ScheduledExecutorService scheduler;
    private final Semaphore inFlight;
    private final Thread dispatcherThread;
    private volatile boolean running;

    private Writer(WriterConfig config, TurbineWebSocketClient client, BlockingQueue<byte[]> responseQueue) {
        this.writerId = UUID.randomUUID();
        this.config = config;
        this.client = client;
        this.responseQueue = responseQueue;
        this.pendingResponses = new ConcurrentHashMap<>();
        this.scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "turbine-writer-scheduler");
            t.setDaemon(true);
            return t;
        });
        this.inFlight = new Semaphore(Math.max(1, config.getMaxInFlight()));
        this.running = true;
        this.dispatcherThread = new Thread(this::dispatchLoop, "turbine-writer-dispatcher");
        this.dispatcherThread.setDaemon(true);
    }

    public static Writer connect(String url) throws TurbineException {
        return connect(new WriterConfig().url(url));
    }

    public static Writer connect(WriterConfig config) throws TurbineException {
        try {
            URI uri = new URI(config.getUrl());
            BlockingQueue<byte[]> responseQueue = new LinkedBlockingQueue<>(RESPONSE_QUEUE_CAPACITY);
            TurbineWebSocketClient client = new TurbineWebSocketClient(uri, responseQueue);

            if (!client.connectBlocking(config.getTimeout().toMillis(), TimeUnit.MILLISECONDS)) {
                throw new TurbineException.ConnectionException("Failed to connect to " + config.getUrl());
            }

            Writer writer = new Writer(config, client, responseQueue);
            if (config.getApiKey() != null) {
                writer.authenticate(config.getApiKey());
            }
            writer.startDispatcher();
            return writer;
        } catch (Exception e) {
            if (e instanceof TurbineException) {
                throw (TurbineException) e;
            }
            throw new TurbineException.ConnectionException("Failed to connect", e);
        }
    }

    private void startDispatcher() {
        dispatcherThread.start();
    }

    private static ByteString uuidToByteString(UUID id) {
        ByteBuffer bb = ByteBuffer.allocate(16);
        bb.putLong(id.getMostSignificantBits());
        bb.putLong(id.getLeastSignificantBits());
        return ByteString.copyFrom(bb.array());
    }

    private void authenticate(String apiKey) throws TurbineException {
        ClientMessage authMessage = ClientMessage.newBuilder()
                .setAuth(AuthRequest.newBuilder().setApiKey(apiKey).build())
                .build();
        client.send(authMessage.toByteArray());

        try {
            byte[] response = responseQueue.poll(config.getTimeout().toMillis(), TimeUnit.MILLISECONDS);
            if (response == null) {
                throw new TurbineException.TimeoutException("Authentication timeout");
            }

            ServerMessage envelope = ServerMessage.parseFrom(response);
            if (envelope.getMessageCase() != ServerMessage.MessageCase.AUTH) {
                throw new TurbineException.ProtocolException("Unexpected auth response");
            }

            io.turbine.sdk.proto.AuthResponse auth = envelope.getAuth();
            if (!auth.getSuccess()) {
                throw new TurbineException.AuthenticationException(
                        "Auth failed (" + auth.getErrorCode() + "): " + auth.getErrorMessage()
                );
            }
            log.debug("Authentication successful");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TurbineException.ConnectionException("Interrupted during auth", e);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
            throw new TurbineException.ProtocolException("Failed to decode auth response", e);
        }
    }

    private void dispatchLoop() {
        while (running) {
            try {
                byte[] response = responseQueue.poll(100, TimeUnit.MILLISECONDS);
                if (response == null) {
                    continue;
                }

                ServerMessage envelope = ServerMessage.parseFrom(response);
                if (envelope.getMessageCase() != ServerMessage.MessageCase.APPEND) {
                    continue;
                }

                AppendResponse resp = envelope.getAppend();
                CompletableFuture<AppendResponse> pending = pendingResponses.remove(resp.getAppendSeq());
                if (pending != null) {
                    pending.complete(resp);
                }
            } catch (InterruptedException e) {
                if (!running) {
                    break;
                }
                Thread.currentThread().interrupt();
            } catch (Exception e) {
                log.warn("Failed to decode writer response", e);
            }
        }
    }

    public UUID getWriterId() {
        return writerId;
    }

    public BatchAck send(int topicId, int partitionId, int schemaId, List<Record> records) throws TurbineException {
        RecordBatch batch = RecordBatch.newBuilder()
                .setTopicId(topicId)
                .setPartitionId(partitionId)
                .setSchemaId(schemaId)
                .addAllRecords(records)
                .build();
        List<BatchAck> acks = sendBatch(List.of(batch));
        if (acks.isEmpty()) {
            throw new TurbineException.ProtocolException("No acks received");
        }
        return acks.get(0);
    }

    public CompletableFuture<BatchAck> sendAsync(
            int topicId,
            int partitionId,
            int schemaId,
            List<Record> records
    ) {
        RecordBatch batch = RecordBatch.newBuilder()
                .setTopicId(topicId)
                .setPartitionId(partitionId)
                .setSchemaId(schemaId)
                .addAllRecords(records)
                .build();
        return sendBatchAsync(List.of(batch)).thenApply(acks -> {
            if (acks.isEmpty()) {
                throw new CompletionException(new TurbineException.ProtocolException("No acks received"));
            }
            return acks.get(0);
        });
    }

    public BatchAck sendOne(int topicId, int partitionId, int schemaId, byte[] key, byte[] value)
            throws TurbineException {
        Record.Builder recordBuilder = Record.newBuilder().setValue(ByteString.copyFrom(value));
        if (key != null) {
            recordBuilder.setKey(ByteString.copyFrom(key));
        }
        return send(topicId, partitionId, schemaId, List.of(recordBuilder.build()));
    }

    public List<BatchAck> sendBatch(List<RecordBatch> batches) throws TurbineException {
        CompletableFuture<List<BatchAck>> future = sendBatchAsync(batches);
        try {
            return future.get(awaitBudgetMillis(), TimeUnit.MILLISECONDS);
        } catch (java.util.concurrent.TimeoutException e) {
            future.cancel(true);
            throw new TurbineException.TimeoutException("Request timeout");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TurbineException.ConnectionException("Interrupted", e);
        } catch (ExecutionException e) {
            Throwable cause = unwrapCompletionThrowable(e.getCause());
            if (cause instanceof TurbineException) {
                throw (TurbineException) cause;
            }
            throw new TurbineException.ProtocolException("Send failed", cause);
        }
    }

    public CompletableFuture<List<BatchAck>> sendBatchAsync(List<RecordBatch> batches) {
        long currentSeq = seq.getAndIncrement();
        AppendRequest req = AppendRequest.newBuilder()
                .setWriterId(uuidToByteString(writerId))
                .setAppendSeq(currentSeq)
                .addAllBatches(batches)
                .build();
        CompletableFuture<List<BatchAck>> result = new CompletableFuture<>();
        sendWithRetryAsync(req, 0, config.getInitialBackoff(), result);
        return result;
    }

    private void sendWithRetryAsync(
            AppendRequest req,
            int retries,
            Duration backoff,
            CompletableFuture<List<BatchAck>> result
    ) {
        if (result.isDone()) {
            return;
        }

        CompletableFuture<AppendResponse> responseFuture;
        try {
            responseFuture = sendRequestAsync(req);
        } catch (TurbineException e) {
            result.completeExceptionally(e);
            return;
        }

        responseFuture.whenComplete((resp, err) -> {
            if (err != null) {
                result.completeExceptionally(unwrapCompletionThrowable(err));
                return;
            }

            if (!resp.getSuccess()) {
                if (resp.getErrorCode() == ERR_BACKPRESSURE) {
                    if (retries >= config.getMaxRetries()) {
                        result.completeExceptionally(
                                new TurbineException.BackpressureException(
                                        "Server backpressure after " + retries + " retries"
                                )
                        );
                        return;
                    }

                    Duration nextBackoff = Duration.ofMillis(
                            Math.min(backoff.toMillis() * 2, config.getMaxBackoff().toMillis())
                    );
                    scheduler.schedule(
                            () -> sendWithRetryAsync(req, retries + 1, nextBackoff, result),
                            backoff.toMillis(),
                            TimeUnit.MILLISECONDS
                    );
                    return;
                }

                result.completeExceptionally(
                        new TurbineException.ProtocolException(
                                "Append failed (" + resp.getErrorCode() + "): " + resp.getErrorMessage()
                        )
                );
                return;
            }

            result.complete(resp.getAppendAcksList());
        });
    }

    private CompletableFuture<AppendResponse> sendRequestAsync(AppendRequest req) throws TurbineException {
        if (!running) {
            throw new TurbineException.ConnectionException("Writer is closed");
        }

        try {
            if (!inFlight.tryAcquire(config.getTimeout().toMillis(), TimeUnit.MILLISECONDS)) {
                throw new TurbineException.TimeoutException("Timed out waiting for in-flight capacity");
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new TurbineException.ConnectionException("Interrupted while waiting for in-flight capacity", e);
        }

        long seqNum = req.getAppendSeq();
        CompletableFuture<AppendResponse> future = new CompletableFuture<>();
        pendingResponses.put(seqNum, future);

        ScheduledFuture<?> timeoutTask = scheduler.schedule(() -> {
            CompletableFuture<AppendResponse> pending = pendingResponses.remove(seqNum);
            if (pending != null) {
                pending.completeExceptionally(new TurbineException.TimeoutException("Request timeout"));
            }
        }, config.getTimeout().toMillis(), TimeUnit.MILLISECONDS);

        future.whenComplete((ignored, ignoredErr) -> {
            timeoutTask.cancel(false);
            inFlight.release();
        });

        try {
            client.send(ClientMessage.newBuilder().setAppend(req).build().toByteArray());
        } catch (Exception e) {
            pendingResponses.remove(seqNum);
            future.completeExceptionally(new TurbineException.ConnectionException("Failed to send request", e));
        }

        return future;
    }

    private long awaitBudgetMillis() {
        long timeout = config.getTimeout().toMillis();
        long retries = (long) config.getMaxRetries() + 1L;
        long retryBudget = config.getMaxBackoff().toMillis() * retries;
        return timeout + retryBudget + 1000;
    }

    private static Throwable unwrapCompletionThrowable(Throwable t) {
        if (t instanceof CompletionException && t.getCause() != null) {
            return t.getCause();
        }
        return t;
    }

    private void failPending(TurbineException error) {
        for (Map.Entry<Long, CompletableFuture<AppendResponse>> entry : pendingResponses.entrySet()) {
            entry.getValue().completeExceptionally(error);
        }
        pendingResponses.clear();
    }

    @Override
    public void close() {
        running = false;
        dispatcherThread.interrupt();
        failPending(new TurbineException.ConnectionException("Writer closed"));
        scheduler.shutdownNow();
        client.close();
    }

    /**
     * Internal WebSocket client.
     */
    private static class TurbineWebSocketClient extends WebSocketClient {
        private static final Logger log = LoggerFactory.getLogger(TurbineWebSocketClient.class);
        private final BlockingQueue<byte[]> responseQueue;

        public TurbineWebSocketClient(URI serverUri, BlockingQueue<byte[]> responseQueue) {
            super(serverUri);
            this.responseQueue = responseQueue;
        }

        @Override
        public void onOpen(ServerHandshake handshake) {
            log.debug("WebSocket connected");
        }

        @Override
        public void onMessage(String message) {
            log.warn("Received text message, expected binary");
        }

        @Override
        public void onMessage(ByteBuffer bytes) {
            byte[] data = new byte[bytes.remaining()];
            bytes.get(data);
            if (!responseQueue.offer(data)) {
                log.warn("Writer response queue full, dropping response frame");
            }
        }

        @Override
        public void onClose(int code, String reason, boolean remote) {
            log.debug("WebSocket closed: {} - {}", code, reason);
        }

        @Override
        public void onError(Exception ex) {
            log.error("WebSocket error", ex);
        }
    }
}

File: sdks/python/flourine/writer.py
"""Writer client for sending messages to Flourine."""

import asyncio
import logging
from dataclasses import dataclass
from typing import Optional
from uuid import UUID, uuid4

import websockets
from websockets.asyncio.client import ClientConnection

from flourine.proto import flourine_wire_pb2 as pb

from .exceptions import (
    AuthenticationException,
    BackpressureException,
    ConnectionException,
    ProtocolException,
    TimeoutException,
)

logger = logging.getLogger(__name__)

ERR_BACKPRESSURE = 1004


@dataclass
class WriterConfig:
    """Configuration for the Flourine writer."""

    url: str = "ws://localhost:9000"
    api_key: Optional[str] = None
    max_retries: int = 5
    initial_backoff: float = 0.1  # seconds
    max_backoff: float = 10.0  # seconds
    timeout: float = 30.0  # seconds
    max_in_flight: int = 256


class Writer:
    """Writer client for sending messages to Flourine."""

    def __init__(
        self,
        ws: ClientConnection,
        config: WriterConfig,
        writer_id: UUID,
    ):
        self._ws = ws
        self._config = config
        self._writer_id = writer_id
        self._seq = 1
        self._lock = asyncio.Lock()
        self._in_flight = asyncio.Semaphore(max(1, config.max_in_flight))
        self._pending: dict[int, asyncio.Future[list[pb.BatchAck]]] = {}
        self._reader_task: Optional[asyncio.Task] = None
        self._closed = False

    @classmethod
    async def connect(cls, url: str) -> "Writer":
        return await cls.connect_with_config(WriterConfig(url=url))

    @classmethod
    async def connect_with_config(cls, config: WriterConfig) -> "Writer":
        try:
            ws = await asyncio.wait_for(
                websockets.connect(config.url),
                timeout=config.timeout,
            )
        except asyncio.TimeoutError:
            raise TimeoutException("Connection timeout")
        except Exception as e:
            raise ConnectionException(f"Failed to connect: {e}")

        writer = cls(ws, config, uuid4())
        if config.api_key:
            await writer._authenticate(config.api_key)
        writer._start_reader()
        return writer

    def _start_reader(self) -> None:
        if self._reader_task is None:
            self._reader_task = asyncio.create_task(self._reader_loop())

    async def _authenticate(self, api_key: str) -> None:
        envelope = pb.ClientMessage()
        envelope.auth.api_key = api_key
        await self._ws.send(envelope.SerializeToString())

        try:
            response = await asyncio.wait_for(
                self._ws.recv(),
                timeout=self._config.timeout,
            )
        except asyncio.TimeoutError:
            raise TimeoutException("Authentication timeout")

        if not isinstance(response, bytes):
            raise ProtocolException("Expected binary response")

        server_msg = pb.ServerMessage()
        server_msg.ParseFromString(response)
        if server_msg.WhichOneof("message") != "auth":
            raise ProtocolException("Unexpected auth response")

        auth_resp = server_msg.auth
        if not auth_resp.success:
            raise AuthenticationException(auth_resp.error_message)

        logger.debug("Authentication successful")

    @property
    def writer_id(self) -> UUID:
        return self._writer_id

    async def send(
        self,
        topic_id: int,
        partition_id: int,
        schema_id: int,
        records: list[pb.Record],
    ) -> pb.BatchAck:
        batch = pb.RecordBatch(
            topic_id=topic_id,
            partition_id=partition_id,
            schema_id=schema_id,
            records=records,
        )
        acks = await self.send_batch([batch])
        if not acks:
            raise ProtocolException("No acks received")
        return acks[0]

    async def send_one(
        self,
        topic_id: int,
        partition_id: int,
        schema_id: int,
        key: Optional[bytes],
        value: bytes,
    ) -> pb.BatchAck:
        record = pb.Record(value=value)
        if key is not None:
            record.key = key
        return await self.send(topic_id, partition_id, schema_id, [record])

    async def send_batch(self, batches: list[pb.RecordBatch]) -> list[pb.BatchAck]:
        async with self._lock:
            current_seq = self._seq
            self._seq += 1

        return await self._send_batch_with_seq(current_seq, batches)

    async def _send_batch_with_seq(
        self, current_seq: int, batches: list[pb.RecordBatch]
    ) -> list[pb.BatchAck]:
        retries = 0
        backoff = self._config.initial_backoff

        while True:
            try:
                acks = await self._send_request(current_seq, batches)
            except BackpressureException:
                if retries >= self._config.max_retries:
                    raise BackpressureException(
                        f"Server backpressure after {retries} retries"
                    )

                logger.warning(
                    "Backpressure from server, retrying in %.2fs (attempt %d/%d)",
                    backoff,
                    retries + 1,
                    self._config.max_retries,
                )
                await asyncio.sleep(backoff)
                backoff = min(backoff * 2, self._config.max_backoff)
                retries += 1
                continue

            logger.debug("Sent %d batches, got %d acks", len(batches), len(acks))
            return acks

    def send_batch_async(self, batches: list[pb.RecordBatch]) -> asyncio.Task:
        return asyncio.create_task(self.send_batch(batches))

    def send_async(
        self,
        topic_id: int,
        partition_id: int,
        schema_id: int,
        records: list[pb.Record],
    ) -> asyncio.Task:
        return asyncio.create_task(
            self.send(topic_id, partition_id, schema_id, records)
        )

    async def _send_request(
        self, seq: int, batches: list[pb.RecordBatch]
    ) -> list[pb.BatchAck]:
        if self._closed:
            raise ConnectionException("Writer is closed")

        req = pb.AppendRequest(
            writer_id=self._writer_id.bytes,
            append_seq=seq,
            batches=batches,
        )
        envelope = pb.ClientMessage()
        envelope.append.CopyFrom(req)
        loop = asyncio.get_running_loop()
        response_future: asyncio.Future[list[pb.BatchAck]] = loop.create_future()

        await self._in_flight.acquire()
        self._pending[seq] = response_future
        try:
            await self._ws.send(envelope.SerializeToString())
        except Exception as e:
            self._pending.pop(seq, None)
            self._in_flight.release()
            raise ConnectionException(f"Send failed: {e}")

        try:
            return await asyncio.wait_for(
                response_future,
                timeout=self._config.timeout,
            )
        except asyncio.TimeoutError:
            self._pending.pop(seq, None)
            raise TimeoutException("Request timeout")
        finally:
            self._in_flight.release()

    async def _reader_loop(self) -> None:
        while True:
            try:
                response = await self._ws.recv()
            except Exception as e:
                if not self._closed:
                    self._fail_all_pending(ConnectionException(f"Connection closed: {e}"))
                break

            if not isinstance(response, bytes):
                continue

            server_msg = pb.ServerMessage()
            try:
                server_msg.ParseFromString(response)
            except Exception:
                continue

            if server_msg.WhichOneof("message") != "append":
                continue

            resp = server_msg.append
            future = self._pending.pop(resp.append_seq, None)
            if future is None or future.done():
                continue

            if not resp.success:
                if resp.error_code == ERR_BACKPRESSURE:
                    future.set_exception(BackpressureException("Server backpressure"))
                else:
                    future.set_exception(
                        ProtocolException(
                            f"Server error {resp.error_code}: {resp.error_message}"
                        )
                    )
                continue

            future.set_result(list(resp.append_acks))

    def _fail_all_pending(self, exc: Exception) -> None:
        for seq, future in list(self._pending.items()):
            if not future.done():
                future.set_exception(exc)
            self._pending.pop(seq, None)

    async def close(self) -> None:
        self._closed = True
        if self._reader_task:
            self._reader_task.cancel()
            try:
                await self._reader_task
            except asyncio.CancelledError:
                pass
        self._fail_all_pending(ConnectionException("Writer closed"))
        await self._ws.close()

    async def __aenter__(self) -> "Writer":
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.close()
